File: client-samples/web/StreamKit/Backends/StreamingBackend.js
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * @fileoverview StreamingBackend interface contract.
 *
 * This is the single seam between `StreamSession` and any networking
 * technology. The library ships with a LiveKit implementation (`LiveKitBackend`);
 * to use a different transport (proprietary streaming SDK, custom WebRTC, etc.)
 * implement this interface and pass your instance to `StreamSession`.
 *
 * There is no runtime code in this file — JavaScript has no formal interfaces.
 * The contract is enforced by documentation and duck-typing.
 *
 * Mirror of the Swift `StreamingBackend` protocol.
 *
 * @module StreamKit/Backends/StreamingBackend
 */

// ─────────────────────────────────────────────────────────────────────────────
// Interface definition (JSDoc only)
// ─────────────────────────────────────────────────────────────────────────────

/**
 * @interface StreamingBackend
 *
 * The contract that every networking backend must satisfy.
 *
 * `StreamSession` delegates all network operations to an object implementing
 * this interface, so call-sites never depend on a specific transport.
 *
 * ## Implementing a custom backend
 *
 * ```js
 * import { ConnectionState } from '../ConnectionState.js';
 *
 * class MyBackend {
 *   // ── Event hooks ─────────────────────────────────────────────────────────
 *   // StreamSession assigns these before calling connect().
 *   // Fire them from any context; StreamSession re-dispatches as needed.
 *
 *   /** @type {((state: string) => void) | null} *\/
 *   onConnectionStateChanged = null;
 *
 *   /** @type {((data: ArrayBuffer | Uint8Array) => void) | null} *\/
 *   onDataReceived = null;
 *
 *   // ── Lifecycle ───────────────────────────────────────────────────────────
 *
 *   async connect(sessionConfig) {
 *     // establish connection using sessionConfig.identity, etc.
 *     this.onConnectionStateChanged?.(ConnectionState.CONNECTED);
 *   }
 *
 *   async disconnect() { /* release resources *\/ }
 *
 *   // ── Media ───────────────────────────────────────────────────────────────
 *
 *   async startCamera() { /* begin capture & publish *\/ }
 *   async stopCamera()  { /* stop capture & unpublish *\/ }
 *
 *   // ── Data channel ────────────────────────────────────────────────────────
 *
 *   async send(data, { reliable = true } = {}) {
 *     // publish data; reliable → ordered/guaranteed, !reliable → best-effort
 *   }
 * }
 *
 * // Then:
 * import { StreamSession } from '../StreamSession.js';
 * const session = new StreamSession(new MyBackend());
 * ```
 */

// ─────────────────────────────────────────────────────────────────────────────
// Property / method stubs (JSDoc typedef style)
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Fired when the connection lifecycle state changes.
 *
 * `StreamSession` sets this property before calling `connect()`.
 * Fire it from any execution context; the session re-dispatches as needed.
 *
 * @name StreamingBackend#onConnectionStateChanged
 * @type {((state: import('../ConnectionState.js').ConnectionState) => void) | null}
 */

/**
 * Fired when binary data arrives from the remote end.
 *
 * `StreamSession` sets this property before calling `connect()`.
 *
 * @name StreamingBackend#onDataReceived
 * @type {((topic: string, data: Uint8Array) => void) | null}
 */

/**
 * Establish a connection using the provided session configuration.
 *
 * Network endpoint details (host, port, token) are supplied at construction
 * time via the backend-specific config object (e.g. `LiveKitConfig`).
 *
 * @function
 * @name StreamingBackend#connect
 * @param {import('../Config/SessionConfig.js').SessionConfig} sessionConfig
 *   Room/participant metadata and media settings.
 * @returns {Promise<void>}
 * @throws {import('../StreamError.js').StreamError}
 */

/**
 * Cleanly disconnect and release all resources (tracks, sockets, etc.).
 *
 * Must resolve even if the session was never connected.
 *
 * @function
 * @name StreamingBackend#disconnect
 * @returns {Promise<void>}
 */

/**
 * Begin capturing the local camera and streaming it to remote participants.
 *
 * Must throw `StreamError.cameraRequiresConnection()` if called while not
 * connected.
 *
 * @function
 * @name StreamingBackend#startCamera
 * @returns {Promise<void>}
 * @throws {import('../StreamError.js').StreamError}
 */

/**
 * Stop camera capture and remove the published video track.
 *
 * Must resolve silently if the camera was never started.
 *
 * @function
 * @name StreamingBackend#stopCamera
 * @returns {Promise<void>}
 * @throws {import('../StreamError.js').StreamError}
 */

/**
 * Send binary data to remote participants via the transport's data channel.
 *
 * Keep individual messages under the transport's MTU (15 KB for LiveKit's
 * WebRTC data channel) to avoid fragmentation.
 *
 * @function
 * @name StreamingBackend#send
 * @param {ArrayBuffer | Uint8Array} data - Payload bytes.
 * @param {object}  [opts]
 * @param {boolean} [opts.reliable=true]
 *   `true` for ordered, guaranteed delivery.
 *   `false` for low-latency best-effort delivery.
 * @returns {Promise<void>}
 * @throws {import('../StreamError.js').StreamError}
 */

File: client-samples/android/app/src/main/java/com/nvidia/xrai/streamkitsample/streamkit/backends/StreamingBackend.kt
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.nvidia.xrai.streamkitsample.streamkit.backends

import com.nvidia.xrai.streamkitsample.streamkit.ConnectionState
import com.nvidia.xrai.streamkitsample.streamkit.config.AudioConfig
import com.nvidia.xrai.streamkitsample.streamkit.config.CameraConfig
import com.nvidia.xrai.streamkitsample.streamkit.config.SessionConfig

/**
 * The contract that every networking backend must satisfy.
 *
 * [StreamSession][com.nvidia.xrai.streamkitsample.streamkit.StreamSession] delegates
 * all network operations to an implementation of this interface, keeping
 * call-sites free of transport-specific details.
 *
 * Mirror of Swift `StreamingBackend` protocol and web `StreamingBackend` interface.
 *
 * ## Implementing a custom backend
 *
 * ```kotlin
 * class MyBackend : StreamingBackend {
 *     override var onConnectionStateChanged: ((ConnectionState) -> Unit)? = null
 *     override var onDataReceived: ((topic: String, data: ByteArray) -> Unit)? = null
 *     override var onAgentStatus: ((status: String) -> Unit)? = null
 *
 *     override suspend fun connect(sessionConfig: SessionConfig) {
 *         // establish connection; call onConnectionStateChanged(CONNECTED) when done
 *     }
 *     override suspend fun disconnect() { /* release resources */ }
 *     override suspend fun startAudio(config: AudioConfig) { /* begin capture & publish */ }
 *     override suspend fun stopAudio() { /* stop & unpublish */ }
 *     override suspend fun startCamera(config: CameraConfig) { /* begin capture & publish */ }
 *     override suspend fun stopCamera() { /* stop & unpublish */ }
 *     override suspend fun send(data: ByteArray, reliable: Boolean) { /* publish data */ }
 * }
 * ```
 */
interface StreamingBackend {

    // ── Event hooks assigned by StreamSession before connect() ────────────────

    /** Fired when the connection lifecycle state changes. */
    var onConnectionStateChanged: ((ConnectionState) -> Unit)?

    /**
     * Fired when binary data arrives from remote participants.
     * `topic` identifies the logical channel (empty string when unset).
     */
    var onDataReceived: ((topic: String, data: ByteArray) -> Unit)?

    /**
     * Fired when an agent publishes a status update on the reserved
     * `_agent.status` channel. Never fires for that topic on [onDataReceived].
     *
     * Common values: `"idle"`, `"processing"`.
     */
    var onAgentStatus: ((status: String) -> Unit)?

    // ── Lifecycle ──────────────────────────────────────────────────────────────

    /**
     * Establishes a connection. Suspends until the connection is ready.
     *
     * Does **not** start audio or camera — call [startAudio] and [startCamera]
     * explicitly after [connect] returns.
     *
     * @throws [com.nvidia.xrai.streamkitsample.streamkit.StreamError]
     */
    suspend fun connect(sessionConfig: SessionConfig)

    /**
     * Cleanly disconnects and releases all resources (tracks, sockets, etc.).
     * Must complete even if [connect] was never called.
     */
    suspend fun disconnect()

    // ── Media ──────────────────────────────────────────────────────────────────

    /**
     * Begins microphone capture and publishes an audio track.
     * @throws [com.nvidia.xrai.streamkitsample.streamkit.StreamError.NotConnected]
     */
    suspend fun startAudio(config: AudioConfig)

    /** Stops microphone capture and unpublishes the audio track. */
    suspend fun stopAudio()

    /**
     * Begins camera capture and publishes a video track.
     * @throws [com.nvidia.xrai.streamkitsample.streamkit.StreamError.CameraRequiresConnection]
     */
    suspend fun startCamera(config: CameraConfig)

    /** Stops camera capture and unpublishes the video track. */
    suspend fun stopCamera()

    // ── Data channel ──────────────────────────────────────────────────────────

    /**
     * Sends binary data to remote participants via the transport's data channel.
     *
     * Keep individual messages under 15 KB (LiveKit's WebRTC data-channel MTU).
     *
     * @param data    Payload bytes.
     * @param reliable `true` for ordered, guaranteed delivery (default).
     *                 `false` for low-latency best-effort delivery.
     * @throws [com.nvidia.xrai.streamkitsample.streamkit.StreamError.NotConnected]
     */
    suspend fun send(data: ByteArray, reliable: Boolean = true)
}
